kafka-transport-binding.md

```diff
@@ -0,0 +1,271 @@
+event attributes MUST be mapped to the Kafka message's 
+[header section][Kafka-Message-Header].
+
+### 1.4. Event Formats
+
+Event formats, used with the *stuctured* content mode, define how an event is
+expressed in a particular data format. All implementations of this
+specification MUST support the [JSON event format][JSON-format].
+
+### 1.5. Security
+
+This specification does not introduce any new security features for Kafka, or
+mandate specific existing features to be used.
+
+## 2. Use of CloudEvents Attributes
+
+This specification does not further define any of the [CloudEvents][CE] event
+attributes.
+
+### 2.1. data Attribute
+
+The `data` attribute is assumed to contain opaque application data that is
+encoded as declared by the `contentType` attribute.
+
+An application is free to hold the information in any in-memory representation
+of its choosing, but as the value is transposed into Kafka as defined in this
+specification, core Kafka provides data available as a sequence of bytes.
+
+For instance, if the declared `contentType` is
+`application/json;charset=utf-8`, the expectation is that the `data` attribute
+value is made available as [UTF-8][RFC3629] encoded JSON text.
+
+## 3. Kafka Message Mapping
+
+With Kafka 0.11.0.0 and above, the content mode is chosen by the sender of the 
+event. Protocol usage patterns that might allow solicitation of events using a
+particular content mode might be defined by an application, but are not defined
+here.
+
+The receiver of the event can distinguish between the two content modes by
+inspecting the `cloudEvents_contentType` property of the Kafka message. If the
+value is prefixed with the CloudEvents media type `application/cloudevents`, 
+indicating the use of a known [event format](#14-event-formats), the receiver
+uses *structured* mode, otherwise it defaults to *binary* mode.
+
+If a receiver finds a CloudEvents media type as per the above rule, but with an
+event format that it cannot handle, for instance
+`application/cloudevents+avro`, it MAY still treat the event as binary and
+forward it to another party as-is.
+
+### 3.1. Binary Content Mode
+
+The *binary* content mode accommodates any shape of event data, and allows for
+efficient transfer and without transcoding effort.
+
+#### 3.1.1. Content Type
+
+For the *binary* mode, the header `cloudEvents_contentType` property MUST be
+mapped directly to the CloudEvents `contentType` attribute.
+
+#### 3.1.2. Event Data Encoding
+
+The [`data` attribute](#22-data-attribute) byte-sequence MUST be used as the
+value of the Kafka message.
+
+#### 3.1.3. Metadata Headers
+
+All [CloudEvents][CE] attributes and
+[CloudEvent Attributes Extensions](primer.md#cloudevent-attribute-extensions) 
+with exception of `data` MUST be individually mapped to and from the Header 
+fields in the Kafka message.
+
+##### 3.1.3.1 Property Names
+
+Cloud Event attributes are prefixed with "cloudEvents_" for use in the
```

Comment:
`cloudEvents_` ? instead of "cloudEvents_" ?

Reply:
I wonder if we should use some sort of fqn:   `io.cloudevents.` to avoid potential clashes ?

Reply:
We are using cloudEvents_ as it has been generally adopted by other transport bindings. I believe we also need to avoid '.' characters. We should be consistent across all bindings.
@duglin  - your thoughts on this?

Reply:
I like consistency, but I think we need to be consistent with that users of the transport expect. I did a quick search and came across this: https://streamsets.com/documentation/controlhub/3.5.0/help/pdesigner/datacollector/UserGuide/Origins/KConsumer.html
and it talks about things like `ssl.truststore.location`  and `com.sun.security.auth.module.Krb5LoginModule`.  So Kafka does seem to be ok with '.'.
I don't know enough about Kafka to know if FQNs would be more "expected" or not, I'll leave that to others - I'd be ok with either.

Reply:
@clemensv - what is your opinion on this (above)? Do we adopt language packaging/namespace semantics like io.cloudevents or cloudEvents_  It does raise the question about event propagation between transports.

Reply:
Just for the record, Apache Kafka never managed to decide on whether "." or "_" are preferred as separators. Configuration parameters are always "." separated (as the Streamset doc shows), but topics are often namespaced with "_" (for example, the internal "__consumer_offsets" topic). I've definitely seen both used in "the wild" (a fact that was sometimes inconvenient as the project had to deal with converting topic names to metric names...)

Either separator will be acceptable by the Apache Kafka community (or rather, either will be objected to, by different sets of people).